### PR TITLE
Mark Framework Interfaces as API part 3

### LINKED ISSUE #32026: Mark Framework interfaces as API part 3
The following interfaces should be marked as API in 2.4-develop branch

1. lib/internal/Magento/Framework/Async/CancelableDeferredInterface.php
1. lib/internal/Magento/Framework/Async/DeferredInterface.php
1. lib/internal/Magento/Framework/Autoload/AutoloaderInterface.php
1. lib/internal/Magento/Framework/Backup/SourceFileInterface.php
1. lib/internal/Magento/Framework/Cache/ConfigInterface.php
1. lib/internal/Magento/Framework/Cache/StaleCacheNotifierInterface.php
1. lib/internal/Magento/Framework/Code/Generator/CodeGeneratorInterface.php
1. lib/internal/Magento/Framework/Code/Minifier/AdapterInterface.php
1. lib/internal/Magento/Framework/Code/Reader/ClassReaderInterface.php
1. lib/internal/Magento/Framework/Code/ValidatorInterface.php
1. lib/internal/Magento/Framework/Component/ComponentRegistrarInterface.php
1. lib/internal/Magento/Framework/Crontab/CrontabManagerInterface.php
1. lib/internal/Magento/Framework/Crontab/TasksProviderInterface.php
1. lib/internal/Magento/Framework/Css/PreProcessor/ErrorHandlerInterface.php
1. lib/internal/Magento/Framework/DB/DataConverter/DataConverterInterface.php
1. lib/internal/Magento/Framework/DB/LoggerInterface.php
1. lib/internal/Magento/Framework/DB/MapperInterface.php
1. lib/internal/Magento/Framework/DB/Query/BatchIteratorInterface.php
1. lib/internal/Magento/Framework/DB/QueryInterface.php
1. lib/internal/Magento/Framework/DB/Select/QueryModifierInterface.php
1. lib/internal/Magento/Framework/DB/Select/RendererInterface.php
1. lib/internal/Magento/Framework/DB/Sequence/SequenceInterface.php
1. lib/internal/Magento/Framework/DB/Sql/ExpressionInterface.php
1. lib/internal/Magento/Framework/Data/Collection/Db/FetchStrategyInterface.php
1. lib/internal/Magento/Framework/Data/Collection/EntityFactoryInterface.php
1. lib/internal/Magento/Framework/Data/CollectionDataSourceInterface.php
1. lib/internal/Magento/Framework/Data/CollectionModifierInterface.php
1. lib/internal/Magento/Framework/Data/SearchResultInterface.php
1. lib/internal/Magento/Framework/Data/SearchResultProcessorInterface.php
1. lib/internal/Magento/Framework/Data/ValueSourceInterface.php
1. lib/internal/Magento/Framework/DataObject/IdentityGeneratorInterface.php
1. lib/internal/Magento/Framework/DataObject/IdentityInterface.php
1. lib/internal/Magento/Framework/DataObject/IdentityValidatorInterface.php
1. lib/internal/Magento/Framework/DataObject/KeyValueObjectInterface.php

### PR DESCRIPTION
**Pull request by mrtuvn**
### Description (*)

The following interfaces marked as API

lib/internal/Magento/Framework/Async/CancelableDeferredInterface.php
lib/internal/Magento/Framework/Async/DeferredInterface.php
lib/internal/Magento/Framework/Autoload/AutoloaderInterface.php
lib/internal/Magento/Framework/Backup/SourceFileInterface.php
lib/internal/Magento/Framework/Cache/ConfigInterface.php
lib/internal/Magento/Framework/Cache/StaleCacheNotifierInterface.php
lib/internal/Magento/Framework/Code/Generator/CodeGeneratorInterface.php
lib/internal/Magento/Framework/Code/Minifier/AdapterInterface.php
lib/internal/Magento/Framework/Code/Reader/ClassReaderInterface.php
lib/internal/Magento/Framework/Code/ValidatorInterface.php
lib/internal/Magento/Framework/Component/ComponentRegistrarInterface.php
lib/internal/Magento/Framework/Crontab/CrontabManagerInterface.php
lib/internal/Magento/Framework/Crontab/TasksProviderInterface.php
lib/internal/Magento/Framework/Css/PreProcessor/ErrorHandlerInterface.php
lib/internal/Magento/Framework/DB/DataConverter/DataConverterInterface.php
lib/internal/Magento/Framework/DB/LoggerInterface.php
lib/internal/Magento/Framework/DB/MapperInterface.php
lib/internal/Magento/Framework/DB/Query/BatchIteratorInterface.php
lib/internal/Magento/Framework/DB/QueryInterface.php
lib/internal/Magento/Framework/DB/Select/QueryModifierInterface.php
lib/internal/Magento/Framework/DB/Select/RendererInterface.php
lib/internal/Magento/Framework/DB/Sequence/SequenceInterface.php
lib/internal/Magento/Framework/DB/Sql/ExpressionInterface.php
lib/internal/Magento/Framework/Data/Collection/Db/FetchStrategyInterface.php
lib/internal/Magento/Framework/Data/Collection/EntityFactoryInterface.php
lib/internal/Magento/Framework/Data/CollectionDataSourceInterface.php
lib/internal/Magento/Framework/Data/CollectionModifierInterface.php
lib/internal/Magento/Framework/Data/SearchResultInterface.php
lib/internal/Magento/Framework/Data/SearchResultProcessorInterface.php
lib/internal/Magento/Framework/Data/ValueSourceInterface.php
lib/internal/Magento/Framework/DataObject/IdentityGeneratorInterface.php
lib/internal/Magento/Framework/DataObject/IdentityInterface.php
lib/internal/Magento/Framework/DataObject/IdentityValidatorInterface.php
lib/internal/Magento/Framework/DataObject/KeyValueObjectInterface.php

Marked api as references
lib/internal/Magento/Framework/Api/CriteriaInterface (by lib/internal/Magento/Framework/DB/MapperInterface.php)

Should merge this before PR related below to avoid conflicts same file

### Dependent Pull Requests
[magento/magento2ce#32135](https://github.com/magento/magento2/pull/32135)

### Fixed Issues (if relevant)

1. Fixes magento/magento2#32026

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by mrtuvn** [comment]
rebased branch 2.5-develop

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
Public type 'Magento\Framework\DB\MapperInterface' references following non-public types:
Magento\Framework\Api\CriteriaInterface

**Comment by mrtuvn** [comment]
Fixed static tests fail

**Comment by mrtuvn** [comment]
@magento run all tests

**Review by eduard13 — CHANGES_REQUESTED** [review]
@mrtuvn please revert the `api` annotation for `Magento/Framework/Api/CriteriaInterface` as it has been already marked in scope of #32135. Let's add it as related PR.

Additionally, could you please enable the strict typing for the adjusted classes?

**Comment by eduard13** [comment]
@mrtuvn please update the PR description (for some reasons I don’t have the permission), Related Pull Requests section with the PR that I mentioned, so it will be used on building the environment.

**Comment by mrtuvn** [comment]
Added related PR

**Comment by eduard13** [comment]
@magento run Static Tests

**Comment by eduard13** [comment]
@mrtuvn please add it according to the following [doc](https://devdocs.magento.com/contributor-guide/maintainer-handbook.html#related-pull-requests).

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/Autoload/AutoloaderInterface.php
Comment (eduard13) [inline lib/internal/Magento/Framework/Autoload/AutoloaderInterface.php]: @mrtuvn it's not a good idea to specify the returning type, as it will fail all the implementations. Let's only add `declare(strict_types=1)`, as you've already did, and should be enough for now.

Comment (mrtuvn) [reply lib/internal/Magento/Framework/Autoload/AutoloaderInterface.php]: so I don't need add return type after every method right ? Only add strict at first line 
Just a bit confuse about consistency across entire ecosystem. While somewhere I see that type return and somewhere it's omitted

Comment (eduard13) [reply lib/internal/Magento/Framework/Autoload/AutoloaderInterface.php]: > so I don't need add return type after every method right ? Only add strict at first line

Exactly.

Also, looks like linking PR (Related PR) didn't work. Could you please double check that you use exactly the same way? The placeholder should also be there.

image

**Comment by mrtuvn** [comment]
Corrected it @eduard13

**Comment by eduard13** [comment]
@magento run Static Tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
added related PR but seem not solve problem with fail static tests

**Comment by eduard13** [comment]
@mrtuvn most likely the PR wasn't added correctly as related. The related PR should be displayed here if everything is correctly added:
image
One more `CE` row should be displayed, pointing to that PR.

Please try multiple variations of adjusting that section, and check the Build Configuration.

**Comment by mrtuvn** [comment]
@magento run Static Tests

**Comment by mrtuvn** [comment]
@magento run Static Tests

**Comment by mrtuvn** [comment]
@magento run Static Tests

**Comment by mrtuvn** [comment]
@magento run Static Tests

**Comment by mrtuvn** [comment]
CC: @slavvka @sidolov @sivaschenko Any ideas what happen with build configuration ? I got fails static when already set related pull requests (maybe wrong syntax) ?

**Comment by eduard13** [comment]
Hi @mrtuvn, after an internal discussion, let's simply adjust the classes, by marking them as `api`, as you initially did. Sorry for the confusion, but looks like relating PRs from the same repo isn't yet supported.

Thank you.

File: lib/internal/Magento/Framework/DB/QueryInterface.php
Comment (eduard13) [inline lib/internal/Magento/Framework/DB/QueryInterface.php]: I’d also revert such adjustments, because of which the SVC is failing, and requires a separate approval step.

**Comment by mrtuvn** [comment]
@magento run Static Tests

**Comment by mrtuvn** [comment]
@magento run WebAPI Tests, Unit Tests, Semantic Version Checker, Sample Data Tests EE, Sample Data Tests CE, Sample Data Tests B2B, Magento Health Index, Integration Tests, Functional Tests EE, Functional Tests CE, Functional Tests B2B, Database Compare

**Comment by mrtuvn** [comment]
@magento run WebAPI Tests,  Semantic Version Checker, Integration Tests

**Comment by eduard13** [comment]
@magento run Integration Tests

**Comment by mrtuvn** [comment]
@magento run Semantic Version Checker

**Comment by eduard13** [comment]
@mrtuvn as I mentioned above, the SVC is failing because of touching the return annotations of some API methods.

**Comment by mrtuvn** [comment]
Yes i have corrected that! Thanks you for the information
@magento run all tests

**Comment by eduard13** [comment]
@magento run Sample Data Tests EE

**Review by eduard13 — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @eduard13, thank you for the review.
[ENGCOM-8833](https://jira.corp.magento.com/browse/ENGCOM-8833) has been created to process this Pull Request

**Comment by mrtuvn** [comment]
Re-targeted PR branch to 2.4-develop as internal discussed with @gabrieldagama

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Functional Tests CE, Functional Tests B2B

**Comment by gabrieldagama** [comment]
@magento import code to https://github.com/magento-engcom/magento2ce

**Comment by magento-engcom-team** [comment]
@gabrieldagama the branch with code successfully imported into`magento-engcom/magento2ce` repository. Branch name: `imported-magento-magento2-32156`.

**Pull request merged**
